### ISSUE TITLE
Inconsistent checkpoint labeling causes problems when resuming training

### ISSUE THREAD

**Issue by cbarkan1** [thinking-machines-lab/tinker-cookbook#367]
The `batch` field in `checkpoints.jsonl` mis-numbers the final checkpoint, which causes problems when resuming training (such as an incorrect learning rate).

Here's a `checkpoints.jsonl` from a 2-epoch SFT run with 9 batches/epoch and `save_every=3`:

```
 name      epoch  batch
 ────────  ─────  ─────
 000003    0      3
 000006    0      6
 000009    1      0
 000012    1      3
 000015    1      6
 final     2      9     ← Incorrect. Should say "final     2      0" (or perhaps "final     1      9").
```

The numbering of the final checkpoint incorrectly indicates that more than an entire epoch has passed since the previous checkpoint. When training is resumed, this causes the following consequences:

**Consequences**:

1. Skipped epoch: After completing an N-epoch run, resuming with `num_epochs=N+1` trains zero steps, but still writes a new checkpoint with an incremented epoch number:

```
 name      epoch  batch
 ────────  ─────  ─────
 ...
 final     2      9
 final     3      9     ← no training occurred, but epoch incremented. Now it looks like we've trained 4 epochs, but we've actually trained 2.
```

2. 🚨 Wrong learning rate due to inflated step counter: If you work around the above issue by setting `num_epochs=N+2` (which will train for 1 additional epoch), Tinker calculates the step number from the inflated epoch, and the learning rate scheduler uses this wrong step number, causing erroneously low learning rate.

Additional issue: Resuming appends a new entry named `"final"` to `checkpoints.jsonl`, leading to non-unique names. Perhaps these could be indexed like `"final_1"`, `"final_2"`, etc.?
